src/Razor/Microsoft.AspNetCore.Razor.Language/src/Extensions/ViewCssScopePass.cs

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Razor.Language.Intermediate;
+
+namespace Microsoft.AspNetCore.Razor.Language.Extensions
+{
+    internal class ViewCssScopePass : IntermediateNodePassBase, IRazorOptimizationPass
+    {
+        // Runs after taghelpers are bound
+        public override int Order => 110;
+
+        protected override void ExecuteCore(RazorCodeDocument codeDocument, DocumentIntermediateNode documentNode)
+        {
+            var cssScope = codeDocument.GetCssScope();
+            if (string.IsNullOrEmpty(cssScope))
+            {
+                return;
+            }
+
+            if (!string.Equals(documentNode.DocumentKind, "mvc.1.0.view", StringComparison.Ordinal) &&
+                !string.Equals(documentNode.DocumentKind, "mvc.1.0.razor-page", StringComparison.Ordinal))
+            {
+                return;
+            }
+
+            var nodes = documentNode.FindDescendantNodes<HtmlContentIntermediateNode>();
+            for (var i = 0; i < nodes.Count; i++)
+            {
+                ProcessElement(nodes[i], cssScope);
+            }
+        }
+
+        private void ProcessElement(HtmlContentIntermediateNode node, string cssScope)
+        {
+            cssScope = " " + cssScope;
```

Comment:
Nitpicky, but instead of allocating a new string for every element, we could construct the space-prefixed version just once and pass it into this method.